consensus/istanbul/uptime/autofix.go

```diff
@@ -0,0 +1,160 @@
+package uptime
+
+import (
+	"math/big"
+
+	"github.com/celo-org/celo-blockchain/consensus/istanbul"
+	"github.com/celo-org/celo-blockchain/core/types"
+)
+
+// autoFixBuilder is an uptime Builder that will fix rewinds and missing headers from a
+// decorated builder.
+type autoFixBuilder struct {
+	builder  Builder
+	provider istanbul.EpochHeadersProvider
+}
+
+func NewAutoFixBuilder(builder Builder, provider istanbul.EpochHeadersProvider) Builder {
+	return &autoFixBuilder{
+		builder:  builder,
+		provider: provider,
+	}
+}
+
+// ProcessHeader ensures the underlying builder is properly build up to the header given.
+// There are a few different cases to manage:
+// 1) Underlying builder is empty. Add all the epoch headers up to the one given
+// 2) New header is behind the builder. Rebuild.
+// 3) New header is the following header, just add.
+// 4) New header is a few headers ahead. Bring missing ones and either add or rebuild if there's a fork
+// 5) New header is the same height as the last one from the builder. Do nothing if it's the same
+// or rebuild if it's a different hash.
+//
+// Note: A big assumption of the autofix is that when given a header, it will be the header returned
+// by the EpochHeadersProvider when calling for epoch headers up to that one. That is, it assumes
+// that it was a valid header stored in the chain.
+func (af *autoFixBuilder) ProcessHeader(header *types.Header) error {
+	number := header.Number.Uint64()
```

Comment:
looks a little weird. We are exposing methods from the builder so that we can do the same checks it does.

I'd imagined this methods as `Try builder.ProcessHeader(), if fails, cleanBuild`. 

Alternatively, if you don't want to have logic on errors, could have a `build.CheckProcesssableHeader() error`

Reply:
This was the way we coded initially, but after discussing it a bit, we went with an approach that doesn't rely on errors as normal code flow for the autofix

Reply:
I guess that works. But we are repeating all the checks we also do in the decoratee. So now you need to remember of modifying them on both sides